src/lib/vasi-macro/src/lib.rs

```diff
@@ -126,50 +149,97 @@ pub fn derive_virtual_address_space_independent(
-            }
-        }
-        syn::Data::Enum(e) => {
-            for variant in &e.variants {
-                for field in &variant.fields {
-                    assertions.push(assertions_for_field(field));
-                }
-            }
-        }
-        syn::Data::Union(u) => {
-            for field in &u.fields.named {
-                assertions.push(assertions_for_field(field));
-            }
-        }
+    // This will contain calls to a function `check` that accepts VirtualAddressSpaceIndependent types,
+    // which is how we validate that the fields are VirtualAddressSpaceIndependent.
+    // e.g. for an input struct definition
+    // ```
+    // struct MyStruct {
+    //   x: u32,
+    //   y: i32,
+    // }
+    // ```
+    //
+    // We'll end up generating code like:
+    // ```
+    // impl VirtualAddressSpaceIndependent for MyStruct {
+    //     const IGNORE: () = {
+    //         let _ = || {
+    //             fn check<T: VirtualAddressSpaceIndependent>() {}
+    //             check::<u32>(); // check type of MyStruct::x
+    //             check::<i32>(); // check type of MyStruct::y
+    //         };
+    //     };
+    // }
+    // ```
+    let types: Vec<&Type> = match &ast.data {
+        syn::Data::Struct(s) => s
+            .fields
+            .iter()
+            .filter(|field| !assume_vasi(&field.attrs))
+            .map(|field| &field.ty)
+            .collect(),
+        syn::Data::Enum(e) => e
+            .variants
+            .iter()
+            .flat_map(|variant| {
+                variant
+                    .fields
+                    .iter()
+                    .filter(|field| !assume_vasi(&field.attrs))
+                    .map(|field| &field.ty)
+            })
+            .collect(),
+        syn::Data::Union(u) => u
+            .fields
+            .named
+            .iter()
+            .filter(|field| !assume_vasi(&field.attrs))
+            .map(|field| &field.ty)
+            .collect(),
     };
-    let assertions: TokenStream = assertions.into_iter().collect();
+
+    // These will fail to compile if any of the types aren't VirtualAddressSpaceIndependent.
+    let calls_to_check: TokenStream = types
+        .into_iter()
+        .map(|ty| quote! {check::<#ty>();})
+        .collect();
+
+    // Add a bound `T: VirtualAddressSpaceIndependent` to every type parameter T.
+    // This allows generic types to be conditionally VirtualAddressSpaceIndependent,
+    // iff their type parameters are.
+    let generics = add_trait_bounds(ast.generics);
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
     let gen = quote! {
-        #assertions
-        /// SAFETY: All fields are vasi::VirtualAddressSpaceIndependent.
-        unsafe impl vasi::VirtualAddressSpaceIndependent for #name {}
+        unsafe impl #impl_generics vasi::VirtualAddressSpaceIndependent for #name #ty_generics #where_clause {
```

Comment:
nit: maybe use `::vasi::VirtualAddressSpaceIndependent` instead so you're always using the `vasi` crate, and not something named `vasi` the user may have defined in their own code.

Reply:
Done